Main 7760 clean

SeanColombo opened the pull request:
Merge all of the refactorings from MAIN-7760. I destroyed the original branch by rebasing (should have been an easy rebase... I'm starting to think that git bash on windows doesn't do rebase correctly).

SeanColombo committed:
Replacing whole commit. MAIN-7760 got obliterated by rebasing again (does rebase just not work correctly in git bash on windows?).

SeanColombo committed:
Changed error-logging to use WikiaLoggger.

WikiaTech commented:
SonarQube analysis reported 4 issues
* ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) 1 minor
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 3 info

Watch the comments in this conversation to review them.

#### 3 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L18](https://github.com/Wikia/app/blob/8e0227c6021436e410546da345bdaa8644644fdc/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L18): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)
1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L21](https://github.com/Wikia/app/blob/8e0227c6021436e410546da345bdaa8644644fdc/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L21): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)
1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L221](https://github.com/Wikia/app/blob/8e0227c6021436e410546da345bdaa8644644fdc/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L221): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)

WikiaTech commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
 		$response = $this->wg->memc->get( $memcKey );
 		if ( empty( $response ) ){
-			$req = MWHttpRequest::factory( self::APP_CONFIG_SERVICE_URL, array( 'noProxy' => true ) );
+			$req = MWHttpRequest::factory( static::APP_CONFIG_SERVICE_URL, array( 'noProxy' => true ) );
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "noProxy" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)

kvas-damian replied:
This block of code related to fetching config looks very similar to `getAllAppConfigs`. Can we reuse it?

SeanColombo replied:
Yup, actually already did that in another branch (see note on line 41). The use-case was identical.

SeanColombo replied:
Will be pulling that change in here in the next commit.

SeanColombo commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
 
 		// Pull in the app-configuration (has data for all apps)
 		$appConfig = [];
 		$memcKey = wfMemcKey( static::$CACHE_KEY, static::$CACHE_KEY_VERSION );
```
In MAIN-7821 branch I've replaced this whole first batch of app-config stuff with a call to getAllAppConfigs() since there's no reason to have this in here twice (I should have replaced this as soon as I created getAllAppConfigs()).

kvas-damian replied:
Data stored in memcache on this key can be shared between wikis. For better performance, let's use `wfSharedMemcKey`:
https://github.com/Wikia/app/blob/64f54f50dcc6b80b296b795c52d63ff476760a52/includes/wikia/GlobalFunctions.php#L691-L691

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
 		wfProfileIn( __METHOD__ );
```
`wfProfileIn()` is deprecated. We already use `xhprof` for profiling so you can simply skip them.

SeanColombo replied:
Ah, w0rd. Removing all calls.

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
-					throw new EmptyResponseException( self::APP_CONFIG_SERVICE_URL );
 				} else {
 					// Request was successful. Cache it in memcached (faster than going over network-card even on our internal network).
 					$this->wg->memc->set( $memcKey, $response, static::$CACHE_EXPIRY );
```
This response have over 200KB as JSON. When we serialize it using php `serialize()` method it gets even bigger. Is there a way to minimize amount of data that we cache here?

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
 		$appConfig = [];
 		$memcKey = wfMemcKey( static::$CACHE_KEY, static::$CACHE_KEY_VERSION );
 		$response = $this->wg->memc->get( $memcKey );
 		if ( empty( $response ) ){
```
You could simplify this code using: https://github.com/Wikia/app/blob/f139fa71cc912c459d364f32e1de40c55c7b950a/includes/wikia/WikiaDataAccess.class.php#L52-L52

See examples: https://github.com/Wikia/app/blob/f139fa71cc912c459d364f32e1de40c55c7b950a/includes/wikia/WikiaDataAccess.class.php#L308-L397

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
 
 		// Since this "Community_App" article won't be found, we need to manually say it's okay so that it's not a 404.
-		$this->response->setCode( self::RESPONSE_OK );
+		$this->response->setCode( static::RESPONSE_OK );
```
Can we use constant that is already defined in `WikiaResponse`?

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -59,7 +63,8 @@ public function index() {
+			$appConfig = [];
 
 		}
 
```
I can't comment line below, but `getConfigForWiki` accepts only one parameter.

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -99,6 +104,7 @@ public function index() {
 	 */
 	public function content( $params ){
 		wfProfileIn( __METHOD__ );
+		$this->debug = "";
```
Is it used anywhere?

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
-			$req = MWHttpRequest::factory( self::APP_CONFIG_SERVICE_URL, array( 'noProxy' => true ) );
+			$req = MWHttpRequest::factory( static::APP_CONFIG_SERVICE_URL, array( 'noProxy' => true ) );
 			$status = $req->execute();
 			if( $status->isOK() ) {
```
What if it isn't ok?

kvas-damian commented:
Line 113: Use `static` keyword

kvas-damian commented:
https://github.com/Wikia/app/blob/50ed4244f881799e77c997e1b4f7f3d7ce5e448f/extensions/wikia/AppPromoLanding/templates/AppPromoLanding_content.php this file deserves for recaftoring.

Templete should include only HTML. Css and JS should be extracted to separate files.

For passing JS variables from controller let's use `addJsConfigVars()` method: https://github.com/Wikia/app/blob/622c22b4d8394e8467ae83d9a429b5cd77eb3029/includes/OutputPage.php#L3013-L3013

kvas-damian commented:
I'm not sure if it helps or not, but you could take different approach for rendering this page. You can find an example on WikiaHubs page: https://github.com/Wikia/app/blob/e6742c247cbe9097271b1dc6a008a0d30c4e337c/extensions/wikia/WikiaHubsV3/hooks/WikiaHubsV3Hooks.php#L10-L10

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -182,13 +191,12 @@ public function content( $params ){
 		$this->trendingArticles = $trendingArticles;
 		$this->mainPageUrl = Title::newMainPage()->getFullUrl();
-		//$this->larrSrc = $this->wg->ExtensionsPath."/wikia/AppPromoLanding/images/arrow-left-long.svg";
 		$this->larrSvgCode = "<svg width=\"22px\" height=\"16px\" viewBox=\"0 0 22 16\" version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">
```
This svg probably could be inlined in scss file.

kvas-damian commented:
you can pass `limit` parameter when fetching top articles:
https://github.com/Wikia/app/blob/349e06a6d505e6cd4722f9fc62068626c5ac1d7e/includes/wikia/api/ArticlesApiController.class.php#L125-L125

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -119,17 +125,20 @@ public function content( $params ){
+			$items = array_slice( $trendingArticlesData, 0, static::MAX_TRENDING_ARTICLES );
 			//load data from response to template
 			foreach( $items as $item ) {
 				$img = $this->app->sendRequest( 'ImageServing', 'getImages', [
```
you can use `ImageServing` class directly. It allows you to fetch data for many articles at once.

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -139,16 +148,16 @@ public function content( $params ){
+		$numThumbsNeeded = ( static::THUMBS_NUM_ROWS * static::THUMBS_PER_ROW );
 		if(count( $trendingArticles ) > $numThumbsNeeded){
 			$trendingArticles = array_slice($trendingArticles, 0, $numThumbsNeeded);
 		} else if(count( $trendingArticles ) < $numThumbsNeeded ){
```
What do you think about using `array_rand` in this condition: http://php.net/manual/en/function.array-rand.php ?

SeanColombo replied:
Since we don't need any help shuffling (afaik) and array_rand gives a warning when trying to pick more than exist in the source-array, I don't think it would help us in this particular instance.

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -99,6 +104,7 @@ public function index() {
 	 */
 	public function content( $params ){
```
This method gets little bit long. Can we think how to split it?

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -72,7 +77,7 @@ public function index() {
 		$srcs = AssetsManager::getInstance()->getGroupCommonURL( 'app_promo_landing_js' );
 		foreach( $srcs as $src ) {
-			$this->wg->Out->addScript( "<script type=\"{$wgJsMimeType}\" src=\"{$src}\"></script>" );
+			$this->wg->Out->addScript( "<script type=\"{$this->wg->JsMimeType}\" src=\"{$src}\"></script>" );
```
What do you think about simply using `addAsset` https://github.com/Wikia/app/blob/2eb1f6e83201550ff4e0be07f02bfdc69c44ee6e/includes/wikia/nirvana/WikiaResponse.class.php#L512-L512 ?

SeanColombo replied:
Cool beans. Will switch to Wikia::addAssetsToOutput() to output the whole group.

kvas-damian commented:
I can see many external requests in this controller, can we think what should happen if any of those fail? How do we handle such situation? At least we should add logging of such case.

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -35,20 +34,25 @@ public function index() {
 				$response = $req->getContent();
 				if ( empty( $response ) ) {
+					// Request failed (app config service failure). Do not cache this response.
+					$this->error(__METHOD__ . ' app config service failure.', [
```
Is it really a failure? We already checked if response is ok, maybe we could make this message more descriptive for future debugging?

kvas-damian commented:
Minor: new array definition syntax is already supported - `[]`

kvas-damian commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -182,13 +191,12 @@ public function content( $params ){
-		//$this->larrSrc = $this->wg->ExtensionsPath."/wikia/AppPromoLanding/images/arrow-left-long.svg";
 		$this->larrSvgCode = "<svg width=\"22px\" height=\"16px\" viewBox=\"0 0 22 16\" version=\"1.1\" xmlns=\"http://www.w3.org/2000/svg\" xmlns:xlink=\"http://www.w3.org/1999/xlink\">
 								<title>BB56E3FE-7480-48C0-96B3-848DAFB20649</title>
 								<desc>Created with sketchtool.</desc>
```
Please use some optimization tools for SVG like https://jakearchibald.github.io/svgomg/

SeanColombo replied:
That tool looks really cool but doesn't appear to be able to handle this SVG... the optimized result doesn't render anything. I trimmed off some obvious cruft though & made the 'title' better.

kvas-damian commented:
@mixth-sense would you like to add anything?

SeanColombo committed:
Did a bunch of tweaks based on kvas-damian review and pulled in a change from MAIN-7821.

WikiaTech commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -211,8 +183,6 @@ public function content( $params ){
 		$this->iosStoreSrc = $this->wg->ExtensionsPath."/wikia/AppPromoLanding/images/appleAppStoreButton.png";
 
 		$this->imgSpacing = 1; // spacing between the image-grid cells.
```
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)

WikiaTech commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -253,19 +221,21 @@ static private function getConfigForWiki( $cityId ){
-					wfProfileOut( __METHOD__ );
-					throw new EmptyResponseException( self::APP_CONFIG_SERVICE_URL );
+					// Request failed (app config service failure). Do not cache this response.
+					$this->error(__METHOD__ . ' app config service failure.', [
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Remove this use of "$this". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2014)

WikiaTech commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -253,19 +221,21 @@ static private function getConfigForWiki( $cityId ){
+					$this->error(__METHOD__ . ' app config service failure.', [
+						'exception' => new Exception()
+					] );
+					$this->response->setCode( \WikiaResponse::RESPONSE_CODE_INTERNAL_SERVER_ERROR );
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Remove this use of "$this". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2014)

WikiaTech commented:
SonarQube analysis reported 5 issues
* ![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) 2 blocker
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 3 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L18](https://github.com/Wikia/app/blob/100274b132e711406ac0ba2f48f2010e48965716/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L18): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)
1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L21](https://github.com/Wikia/app/blob/100274b132e711406ac0ba2f48f2010e48965716/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L21): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)

SeanColombo committed:
More cleanups (called a shorter method for including scripts, saved a few dozen bytes on an inline svg).

WikiaTech commented on extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php:
```diff
@@ -224,11 +190,10 @@ public function content( $params ){
 
 		// Gets the configs for ALL apps.
-		$appConfig = AppPromoLandingController::getAllAppConfigs();
+		$appConfig = $this->getAllAppConfigs();
```
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Remove this use of "$this". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS2014)

WikiaTech commented:
SonarQube analysis reported 4 issues
* ![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) 1 blocker
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 3 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L18](https://github.com/Wikia/app/blob/a455fbd98e4275e618cf5dab53cbf414a4300394/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L18): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)
1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L21](https://github.com/Wikia/app/blob/a455fbd98e4275e618cf5dab53cbf414a4300394/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L21): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)

SeanColombo committed:
Untangled some static vs instance issues that I created in last two commits.

WikiaTech commented:
SonarQube analysis reported 3 issues
* ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 3 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L17](https://github.com/Wikia/app/blob/06f2078b0e0f2ebfe0f448df23433fe06c8a9d51/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L17): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)
1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L20](https://github.com/Wikia/app/blob/06f2078b0e0f2ebfe0f448df23433fe06c8a9d51/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L20): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)

SeanColombo committed:
New array def convention in a couple of spots.

WikiaTech commented:
SonarQube analysis reported 3 issues
- ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) 3 info

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:
1. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L17](https://github.com/Wikia/app/blob/e7e391455d6b07c11805a4e127de2867334271b8/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L17): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)
2. ![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) [AppPromoLandingController.class.php#L20](https://github.com/Wikia/app/blob/e7e391455d6b07c11805a4e127de2867334271b8/extensions/wikia/AppPromoLanding/AppPromoLandingController.class.php#L20): Move this trailing comment on the previous empty line. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS139)

SeanColombo commented:
Still a couple more improvements tweaks could be made, but what we have is already some really useful improvements (and these changes are diverging from MAIN-7821) so gonna merge what we have and then jump back to that branch's task.

I'll be referring back to the un-addressed comments in this branch & continue to tweak based on that. One that I think really should be addressed is the case where branch key isn't in memcached and branch's API is down... right now we'd just do empty keys, but we should do some noisy logging and at least think if there is a better way to handle the page (do we want to show it or totally error-out? just show an error message when user clicks Get App button?)
